### PR TITLE
New: Rule SRI

### LINKED ISSUE #26: Add rule to check subresource integrity
- [x] `<script>` or `<link>` elements don't have the `integrity` attribute 
- [x] the content of the `<script>` or `<link>` elements doesn’t match the associated `integrity` value 
- [x] other checks (TODO: read the spec)

----

See also:

* https://developer.mozilla.org/en-US/docs/Web/Security/Subresource_Integrity
* Spec: https://w3c.github.io/webappsec-subresource-integrity/

### PR DESCRIPTION
**Pull request by molant**
## Pull request checklist

Make sure you:

- [X] Signed the [Contributor License Agreement](https://cla.js.foundation/sonarwhal/sonarwhal)
- [X] Followed the [commit message guidelines](https://sonarwhal.com/docs/contributor-guide/getting-started/pull-requests/#commit-messagess)

For non-trivial changes, please make sure you also:

- [X] Added/Updated related documentation.
- [X] Added/Updated related tests.

## Short description of the change(s)

Close #26

### CODE REVIEW

**Comment by molant** [comment]
The spec talks about `integrity` for `script` and `link` but all the examples are for styles so I'm not sure if it applies also for manifests, etc. Also CSP values are only for scripts and styles.
For now I'm limiting to styles and scripts and ask around.

File: packages/rule-sri/README.md [hunk 1]
```diff
@@ -0,0 +1,145 @@
+# Require scripts and styles to use Subresource integrity (`sri`)
+
+`sri` warns about requesting scripts or styles without using Subresource
+integrity.
+
+## Why is this important?
+
+Nowadays it's very common to use third party resources from CDNs or different
+services (analytics, ads, etc.) and thus increasing the risk surface of your
```
Comment (alrra) [inline packages/rule-sri/README.md]: > , and thus,

File: packages/rule-sri/README.md [hunk 2]
```diff
@@ -0,0 +1,145 @@
...
+(TLS, HSTS, etc.), an attacker (or administrator) with access to the server can
+manipulate the content with impunity.
+
+> If you want to load a crypto miner on 1,000+ websites you don't attack 1,000+
+websites, you attack the 1 website that they all load content from.
+([Scott Helme][weak link])
+
+Subresource integrity [is a standard][sri spec] that mitigates this by ensuring
+that an exact representation of a resource, and only that representation, loads
+and executes.
```
Comment (alrra) [inline packages/rule-sri/README.md]: Maybe we should add a warning that this is only guaranteed over HTTPS?

File: packages/rule-sri/README.md [hunk 3]
```diff
@@ -0,0 +1,145 @@
...
+  contain something in the form of `sha(256|384|512)-HASH`, where `HASH` is
+  the hashed value of the downlaoded body's response using the previous
+  algorithm (`sha256`, `sha384`, or `sha512`).
+* The minium cryptographic hash function used has to be [`sha384`][collisions].
+  If multiple are provided, the highest one will be used to determine if the
+  baseline is met.
+* When using a cross-origin resource (e.g.: using a script hosted in a third
+  party CDN), the `<script>` tag needs to have a valid
+  [`crossorigin` attribute][crossorigin].
+* The resource needs to be served on a [secure context][secure context]
```
Comment (alrra) [inline packages/rule-sri/README.md]: > The resource is served...

File: packages/rule-sri/README.md [hunk 4]
```diff
@@ -0,0 +1,145 @@
...
+  (i.e.: HTTPS)
+* The hash from the `integrity` attribute needs to be the same as the one
+  calculated using the response's body.
+* If multiple hashes are provided, at least one needs to be valid
```
Comment (alrra) [inline packages/rule-sri/README.md]: > valid.

File: packages/rule-sri/README.md [hunk 5]
```diff
@@ -0,0 +1,145 @@
...
+Same-origin resource with no `integrity`:
+
+```html
+<link rel="stylesheet" href="/style.css">
+```
+
+Same-origin resource with hash function less secure than `sha384`:
+
+```html
+<script src="/script.js" integrity="sha256-validHashHere">
```
Comment (alrra) [inline packages/rule-sri/README.md]: Maybe to be consistent we should use a random hash?

File: packages/rule-sri/README.md [hunk 6]
```diff
@@ -0,0 +1,145 @@
...
+
+## Can the rule be configured?
+
+Yes, by default the baseline algorithm is `sha384` but you can change it to
+`sha256`, or `sha512`:
+
+```json
+"sri": ["warning", {
+    "baseline": "sha512"
+}
```
Comment (alrra) [inline packages/rule-sri/README.md]: > }]

File: packages/rule-sri/package.json [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+  "name": "@sonarwhal/rule-sri",
+  "nyc": {
+    "extends": "../../../../.nycrc"
+  },
+  "peerDependencies": {
+    "sonarwhal": "^1.0.0"
+  },
+  "repository": "sonarwhal/sonarwhal",
+  "scripts": {
+    "build": "npm run clean && npm-run-all build:*",
```
Comment (alrra) [inline packages/rule-sri/package.json]: Add `"build-release": "npm run clean && npm run build:assets && tsc --inlineSourceMap false --removeComments true",`

File: packages/rule-sri/package.json [hunk 8]
```diff
@@ -0,0 +1,69 @@
...
+    "sri-rule"
+  ],
+  "license": "Apache-2.0",
+  "main": "dist/src/index.js",
+  "name": "@sonarwhal/rule-sri",
+  "nyc": {
+    "extends": "../../../../.nycrc"
+  },
+  "peerDependencies": {
+    "sonarwhal": "^1.0.0"
```
Comment (alrra) [inline packages/rule-sri/package.json]: > 1.0.1

File: packages/rule-sri/package.json [hunk 9]
```diff
@@ -0,0 +1,69 @@
...
+    "lint:js": "eslint . --cache --ext js --ext md --ext ts --ignore-path ../../.eslintignore --report-unused-disable-directives",
+    "lint:md": "markdownlint *.md",
+    "sonarwhal": "node node_modules/sonarwhal/dist/src/bin/sonarwhal.js",
+    "test": "npm run lint && npm run build && nyc ava",
+    "watch": "npm run build && npm-run-all --parallel -c watch:*",
+    "watch:assets": "npm run build:assets -- -w --no-initial",
+    "watch:test": "ava --watch",
+    "watch:ts": "npm run build:ts -- --watch"
+  },
+  "version": "0.1.0"
```
Comment (alrra) [inline packages/rule-sri/package.json]: Add `"private": true`

File: packages/rule-sri/package.json [hunk 10]
```diff
@@ -0,0 +1,69 @@
+{
+  "author": "",
```
Comment (alrra) [inline packages/rule-sri/package.json]: Remove this.

File: packages/rule-sri/package.json (hunk 9 above)
Comment (alrra) [inline packages/rule-sri/package.json]: > 1.0.0

File: packages/rule-sri/package.json [hunk 11]
```diff
@@ -0,0 +1,69 @@
...
+  },
+  "description": "Require scripts and styles to use Subresource Integrity",
+  "devDependencies": {
+    "ava": "^0.25.0",
+    "cpx": "^1.5.0",
+    "eslint": "^4.18.2",
+    "eslint-plugin-markdown": "^1.0.0-beta.7",
+    "eslint-plugin-typescript": "^0.10.0",
+    "markdownlint-cli": "^0.7.0",
+    "npm-link-check": "^2.0.0",
```
Comment (alrra) [inline packages/rule-sri/package.json]: I think this ca be removed./

File: packages/rule-sri/src/index.ts [hunk 12]
```diff
@@ -0,0 +1,7 @@
+/**
+ * @fileoverview Require scripts and styles to use Subresource Integrity
+ */
+
+import * as sri from './sri';
```
Comment (alrra) [inline packages/rule-sri/src/index.ts]: Maybe we should just make a convention that if there is a single rule in a package we use `rule.ts` for the name of the file?

File: packages/rule-sri/src/sri.ts [hunk 13]
```diff
@@ -0,0 +1,305 @@
...
+
+import * as crypto from 'crypto';
+import { URL } from 'url';
+import { promisify } from 'util';
+
+import * as async from 'async';
+
+import { Category } from 'sonarwhal/dist/src/lib/enums/category';
+import { RuleContext } from 'sonarwhal/dist/src/lib/rule-context';
+// The list of types depends on the events you want to capture.
```
Comment (alrra) [inline packages/rule-sri/src/sri.ts]: Remove this line.

File: packages/rule-sri/src/sri.ts [hunk 14]
```diff
@@ -0,0 +1,305 @@
...
+     * * `same-origin`
+     * * `cross-origin` on a CORS-enabled request
+     *
+     * More info in https://w3c.github.io/webappsec-subresource-integrity/#is-response-eligible
+     */
+    private async isEligibleForIntegrityValidation(evt: FetchEnd): Promise<boolean> {
+        debug('Is eligible for integrity validation?');
+
+        const { element, resource } = evt;
+        const resourceOrigin: string = new URL(resource).origin;
```
Comment (alrra) [inline packages/rule-sri/src/sri.ts]: Maybe use [same-origin](https://github.com/sonarwhal/sonarwhal/blob/master/packages/rule-disown-opener/package.json#L12) as things may be more complex then this?

File: packages/rule-sri/src/sri.ts [hunk 15]
```diff
@@ -0,0 +1,305 @@
...
+
+        // CORS validation only applies to scripts, styles are OK
```
Comment (alrra) [inline packages/rule-sri/src/sri.ts]: Why are styles ok? Maybe provide a link?

File: packages/rule-sri/src/sri.ts [hunk 16]
```diff
@@ -0,0 +1,305 @@
...
+        // cross-origin scripts need to be loaded with a valid "crossorigin" attribute (ie.: anonymous or use-credentials)
+        const crossorigin = element.getAttribute('crossorigin');
+
+        if (!crossorigin) {
+            await this.context.report(resource, element, `Cross-origin scripts need a "crossorigin" attribute to be elegible for integrity validation`);
+
+            return false;
+        }
+
+        const validCrossorigin = crossorigin === 'anonymous' || crossorigin === 'use-credentials';
```
Comment (alrra) [inline packages/rule-sri/src/sri.ts]: What if I have `crossorigin="anonymous use-credentials"`? Don't you need to parse the value for it, and then check it it contains a certain value?

File: packages/rule-sri/src/sri.ts [hunk 17]
```diff
@@ -0,0 +1,305 @@
...
+     *              sha384-+/M6kredJcxdsqkczBUjMLvqyHb1K/JThDXWsBVxMEeZHEaMKEOEct339VItX1zB"
+     *   crossorigin="anonymous"></script>`
+     *
+     * https://w3c.github.io/webappsec-subresource-integrity/#agility
+     */
+    private async isIntegrityFormatValid(evt: FetchEnd): Promise<boolean> {
+        debug('Is integrity attribute valid?');
+        const { element, resource } = evt;
+        const integrity = element.getAttribute('integrity');
+        const integrityRegExp = /^sha(256|384|512)-/;
```
Comment (alrra) [inline packages/rule-sri/src/sri.ts]: What if I have multiple `integrity` attributes?

File: packages/rule-sri/src/sri.ts [hunk 18]
```diff
@@ -0,0 +1,305 @@
...
+        if (element.nodeName !== 'SCRIPT') {
+            return true;
+        }
+
+        if (this.origin === resourceOrigin) {
+            return true;
+        }
+
+        // cross-origin scripts need to be loaded with a valid "crossorigin" attribute (ie.: anonymous or use-credentials)
+        const crossorigin = element.getAttribute('crossorigin');
```
Comment (alrra) [inline packages/rule-sri/src/sri.ts]: Use normalizeString?

File: packages/rule-sri/README.md (hunk 5 above)
Comment (molant) [reply packages/rule-sri/README.md]: I was thinking about that, but I though it was clearer this way so users reading about SRI for the first time know they need to calculate the hash and replace it.

Comment (alrra) [reply packages/rule-sri/README.md]: ok.

File: packages/rule-sri/src/sri.ts (hunk 14 above)
Comment (molant) [reply packages/rule-sri/src/sri.ts]: I actually think this code is better than [the code in that package](https://github.com/rexxars/same-origin/blob/master/index.js). It's using the old `url.parse` and doing some magic, and we get for free the `origin` if we use `new URL`. Maybe we could add a method in `misc` so we can remove that dependency from the project.

File: packages/rule-sri/src/sri.ts (hunk 15 above)
Comment (molant) [reply packages/rule-sri/src/sri.ts]: That's the link that's just above in the method description

File: packages/rule-sri/src/sri.ts (hunk 14 above)
Comment (alrra) [reply packages/rule-sri/src/sri.ts]: > Maybe we could add a method in misc so we can remove that dependency from the project.

:+1:

File: packages/rule-sri/src/sri.ts (hunk 15 above)
Comment (alrra) [reply packages/rule-sri/src/sri.ts]: OK.

File: packages/rule-sri/README.md (hunk 2 above)
Comment (molant) [reply packages/rule-sri/README.md]: I put a comment in the latest comment in the next paragraph

File: packages/rule-sri/src/sri.ts (hunk 18 above)
Comment (molant) [reply packages/rule-sri/src/sri.ts]: Done

File: packages/rule-sri/src/sri.ts (hunk 17 above)
Comment (molant) [reply packages/rule-sri/src/sri.ts]: I've added another test. Only the first instance is accessible via `getAttribute`. Additionally, it looks like Chrome tries to download the file twice if the first hash is invalid, but only the first attribute is accessible. I've left the code of the test commented with an explanation.

File: packages/rule-sri/src/sri.ts (hunk 14 above)
Comment (molant) [reply packages/rule-sri/src/sri.ts]: I've created https://github.com/sonarwhal/sonarwhal/issues/879 for this

File: packages/rule-sri/src/sri.ts (hunk 13 above)
Comment (molant) [reply packages/rule-sri/src/sri.ts]: Done

File: packages/rule-sri/src/index.ts (hunk 12 above)
Comment (molant) [reply packages/rule-sri/src/index.ts]: Done

File: packages/rule-sri/package.json (hunk 11 above)
Comment (molant) [reply packages/rule-sri/package.json]: Done

File: packages/rule-sri/package.json (hunk 9 above)
Comment (molant) [reply packages/rule-sri/package.json]: Done

File: packages/rule-sri/package.json (hunk 10 above)
Comment (molant) [reply packages/rule-sri/package.json]: Done

File: packages/rule-sri/package.json (hunk 9 above)
Comment (molant) [reply packages/rule-sri/package.json]: Done

File: packages/rule-sri/package.json (hunk 8 above)
Comment (molant) [reply packages/rule-sri/package.json]: Actually `1.0.3`, things move fast!

File: packages/rule-sri/package.json (hunk 7 above)
Comment (molant) [reply packages/rule-sri/package.json]: Done

File: packages/rule-sri/README.md (hunk 6 above)
Comment (molant) [reply packages/rule-sri/README.md]: Done

File: packages/rule-sri/README.md (hunk 4 above)
Comment (molant) [reply packages/rule-sri/README.md]: Done

File: packages/rule-sri/README.md (hunk 3 above)
Comment (molant) [reply packages/rule-sri/README.md]: Done

**Comment by molant** [comment]
@alrra all your feedback should be included.

@sonarwhal/core this is ready for another round.

**Comment by alrra** [comment]
Also, maybe add this to `configuration-web-recommended`?

**Comment by molant** [comment]
👍

________________________________

File: packages/rule-sri/src/rule.ts [hunk 19]
```diff
@@ -0,0 +1,313 @@
...
+    private origin: string;
+    private baseline: string = 'sha384';
+
+    /**
+     * Returns the hash of the content for the given `sha` strengh in a format
+     * valid with SRI:
+     * * base64
+     * * `sha384-hash`
+     */
+    private calculateHash(content: string, sha): string {
```
Comment (alrra) [inline packages/rule-sri/src/rule.ts]: > sha: string

File: packages/rule-sri/src/rule.ts [hunk 20]
```diff
@@ -0,0 +1,313 @@
...
+        if (!element) {
+            return Promise.resolve(false);
+        }
+
+        /*
+         * The element is not one that we care about (could be an img, video, etc.)
+         * No need to report anything, but we can stop processing things right away.
+         */
+        const isScript: boolean = element.nodeName === 'SCRIPT' && !!element.getAttribute('src');
+        const isStyle: boolean = element.nodeName === 'LINK' && element.getAttribute('rel') === 'stylesheet';
```
Comment (alrra) [inline packages/rule-sri/src/rule.ts]: `normalizeString(element.getAttribute('rel')) === 'stylesheet';`

File: packages/rule-sri/LICENSE.txt [hunk 21]
```diff
@@ -0,0 +1,201 @@
+                                 Apache License
```
[the remark this answers is not in the document]
Comment (molant) [reply packages/rule-sri/LICENSE.txt]: It's taken care by the generators. I wouldn't change anything TBH.

[the remark this answers is not in the document]
Comment (alrra) [reply packages/rule-sri/LICENSE.txt]: OK.

File: packages/rule-sri/README.md [hunk 22]
```diff
@@ -0,0 +1,145 @@
+# Require scripts and styles to use subresource integrity (`sri`)
+
+`sri` warns about requesting scripts or styles without using Subresource
+integrity.
+
+## Why is this important?
+
+Nowadays it's very common to use third party resources from CDNs or different
+services (analytics, ads, etc.), and thus, increasing the risk surface of your
```
[the remark this answers is not in the document]
Comment (molant) [reply packages/rule-sri/README.md]: I'm pretty sure it's `increasing`

**Comment by molant** [comment]
Feedback is pushed

**Pull request closed without merging**